core/controllers/editor_test.py

```diff
@@ -605,6 +587,332 @@ def test_state_yaml_handler(self):
+        self.login(self.OWNER_EMAIL)
+        exp_id = 'eid'
+        owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+
+        exploration = self.save_new_valid_exploration(exp_id, owner_id)
+
+        answers = stats_services.get_top_state_answer_stats_multi(
+            exp_id, exploration.states)
+
+        response = self.get_json(
+            '/createhandler/state_answer_stats/%s' % exp_id)
+
+        self.assertEqual(response['answers'], answers)
+
+        self.logout()
+
+
+class TopUnresolvedAnswersHandlerTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(TopUnresolvedAnswersHandlerTests, self).setUp()
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+
+        self.exp_id = 'exp_id'
+        self.exploration = self.save_new_valid_exploration(
+            self.exp_id, self.owner_id)
+
+    def test_cannot_get_unresolved_answers_with_no_state_name(self):
+        self.login(self.OWNER_EMAIL)
+
+
+        self.get_json(
+            '/createhandler/get_top_unresolved_answers/%s' % self.exp_id,
+            expected_status_int=404)
+
+        self.logout()
+
+    def test_get_top_unresolved_answers(self):
+        self.login(self.OWNER_EMAIL)
+
+        answers = stats_services.get_top_state_unresolved_answers(
+            self.exp_id, self.exploration.init_state_name)
+
+        response = self.get_json(
+            '/createhandler/get_top_unresolved_answers/%s?state_name=%s'
+            % (self.exp_id, self.exploration.init_state_name))
+
+        self.assertEqual(response['unresolved_answers'], answers)
+
+        self.logout()
+
+
+class StateRulesStatsHandlerTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(StateRulesStatsHandlerTests, self).setUp()
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+
+    def test_get_with_invalid_exploration_id_raises_error(self):
+        self.login(self.OWNER_EMAIL)
+
+        self.get_json(
+            '/createhandler/state_rules_stats/invalid_exp_id/state_name',
+            expected_status_int=404)
+
+        self.logout()
+
+    def test_cannot_get_learner_answer_statistics_with_invalid_state_name(
+            self):
+        observed_log_messages = []
+
+        def _mock_logging_function(msg, *args):
+            """Mocks logging.error()."""
+            observed_log_messages.append(msg % args)
+
+        logging_swap = self.swap(logging, 'error', _mock_logging_function)
+
+        self.login(self.OWNER_EMAIL)
+        exp_id = 'eid'
+        owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+
+        self.save_new_valid_exploration(exp_id, owner_id)
+
+        with logging_swap:
+            self.get_json(
+                '/createhandler/state_rules_stats/%s/invalid_state_name'
+                % (exp_id), expected_status_int=404)
+
+        self.assertEqual(len(observed_log_messages), 2)
```

Comment:
This and the following lines read oddly. Why are we only checking one of these?

Reply:
Done